### ISSUE TITLE
pyorient_native is causing a memory leak in pyorient.

### ISSUE THREAD

**Issue by anber500** [nikulukani/pyorient_native#1]
I'm trying to make use of the Binary Serialization for pyorient. The Binary Serializor is the only method that currently works for OrientDB in distributed mode.

Although everything seems to work, there is a very bad memory leak in pyorient when using the Binary deserializor. The memory leak seem to happen in pyorient_native.

I've tracked the issue down to this line of code in pyorient:

```
clsname, data = pyorient_native.deserialize(content,
                                content.__sizeof__(), self.props)
```

I'm not very familiar with C++ so I can't fix it myself :(

Except for the Memory leak, I'm loving the Binary deserializor so far :)

**Comment by anber500** [comment]
I have been looking into the code myself.

```
static PyObject* native_deserialize(PyObject *self, PyObject *args){
  PyObject * pycontent;
  int len;
  PyObject * props  = NULL;
  PyArg_ParseTuple(args, "|SiO", &pycontent, &len, &props);

  RecordParser reader("onet_ser_v0");
  TrackerListener* listener;

  listener = new TrackerListener(props);
  #if PY_MAJOR_VERSION >= 3
  reader.parse((unsigned char*)PyBytes_AsString(pycontent), len, *listener);
  #else
  reader.parse((unsigned char*)PyString_AsString(pycontent), len, *listener);
  #endif
  return listener->obj;
}
```

The destructor for the listener never gets called. If I change the interface to the following, then the destructor is called when I delete the listener:

```
class TrackerListener: public RecordParseListener {
 public:
 PyObject *obj;
 string *cur_field;
  stack<OType> types_stack;
  stack<PyObject*> obj_stack;

  virtual void startDocument(const char * name,size_t name_length) ;
....
  TrackerListener(PyObject* props) ;
  virtual ~TrackerListener() ;
};
```

The other issue seems to be that "native_deserialize" should not return the obj in the listener directly. Currently the destructor never gets called. Can't we pass a pointer to the listener instead? That way we can cleanup the listener before returning?

**Comment by nikulukani** [comment]
Thanks for bringing this issue up. The main reason for the memory leaks is incorrect handling of reference counts for python objects (https://docs.python.org/2/c-api/intro.html#reference-counts). I fixed it at places that were obvious to me upon a first pass-through in the latest commit. I haven't had a chance to test it thoroughly yet. Let me know if the latest commit helps, I will be spending some time on this issue later in the week to see if I missed any other points in the code that might be causing a memory leak.

**Comment by anber500** [comment]
cool thankx. I'll test it.

**Comment by anber500** [comment]
The problem is slightly better but there are still leaks. I'll see if i can spot something myself.

**Comment by nikulukani** [comment]
I just ran some tests for deserializations and there were no leaks for the queries I ran. Did the leaks occur for you during serialization or deserialization?

**Comment by nikulukani** [comment]
I ask as I haven't gone through encoder.cpp yet but will fix the leaks there by tonight.

**Comment by anber500** [comment]
I pull the latest changes and it seems like the bug has been fixed. Thanks so much. Joy!

**Comment by nikulukani** [comment]
Glad to hear that! I will run some more tests just to make sure there are no lingering memory leaks. Will bump the version and push to pypi at that point. 

If you are using the binary serialization, I suggest to apply the patch in pull request #216 on pyorient (https://github.com/mogui/pyorient/pull/216). Without that, you might get some unexpected results.

**Comment by anber500** [comment]
Thanks, I'll do that.
I hope the binary serializer makes it into the next pyorient release because as it stands, pyorient's CSV serializer doesn't support OrientDB in distributed mode and I really need that.
I forked your code for my dev environment but I could much rather wait for the pypi because I have to pull it into my docker build as well. (which is where I noticed the problem in the first place)
